DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp

```diff
@@ -0,0 +1,170 @@
+            << q.get_device().get_info<sycl::info::device::name>() << std::endl;
+
+  try {
+    // Set up buffers
+    buffer imgplot_buf((rgb*)image_plot,
+                       range(img_dimensions * img_dimensions));
+    buffer coords_buf((coordinate*)coords, range(size_n));
+    buffer total_buf((int*)(&total), range(1));
+
+    // Perform Monte Carlo simulation and reduce results
+    q.submit([&](handler& h) {
+      // Set up accessors
+      auto imgplot_acc = imgplot_buf.get_access<access::mode::write>(h);
+      auto coords_acc = coords_buf.get_access<access::mode::read_write>(h);
+      auto total_acc = total_buf.get_access<access::mode::read_write>(h);
+
+      // Monte Carlo Procedure + Reduction
+      h.parallel_for(nd_range<1>(num_wg * size_wg, size_wg),
+                     sycl::intel::reduction(total_acc, 0, std::plus<int>()),
+                     [=](nd_item<1> it, auto& total_acc) {
+                       int i =
+                           it.get_global_id();  // Index for accessing buffers
+
+                       if (i < size_n) {  // only runs if a work item's ID has a
+                                          // corresponding sample coordinate
+                         // Get random coords
+                         double x = coords_acc[i].x;
+                         double y = coords_acc[i].y;
+
+                         // Check if coordinates are bounded by a circle of
+                         // radius 1
+                         double hypotenuse_sqr = (x * x + y * y);
+                         if (hypotenuse_sqr <= 1.0) {  // If bounded
+                           // increment total
+                           total_acc += 1;
+                           // Draw sample point in image plot
+                           imgplot_acc[GetPixelIndex(x, y)].red = 0;
+                           imgplot_acc[GetPixelIndex(x, y)].green = 255;
+                           imgplot_acc[GetPixelIndex(x, y)].blue = 0;
+                         } else {
+                           // Draw sample point in image plot
+                           imgplot_acc[GetPixelIndex(x, y)].red = 255;
+                           imgplot_acc[GetPixelIndex(x, y)].green = 0;
+                           imgplot_acc[GetPixelIndex(x, y)].blue = 0;
+                         }
+                       }
+                     });
+    });
+    q.wait_and_throw();
+
+  } catch (sycl::exception e) {
+    std::cout << "SYCL exception caught: " << e.what() << std::endl;
+    exit(1);
+  }
+
+  // return calculated value of pi
+  return 4.0 * (double)total / size_n;
+}
+
+int main() {
+  // Validate constants
+  if (size_n < size_wg) {
+    std::cout << "ERROR: size_n must be greater than or equal to size_wg\n"
+              << std::endl;
+    exit(1);
+  }
+
+  // Initialize random seed
+  srand(time(NULL));
+
+  // Allocate memory for the output image
+  rgb* image_plot = (rgb*)calloc(img_dimensions * img_dimensions, sizeof(rgb));
```

Comment:
Bad coding style:
- spaghetti C
- no check for bad alloc
- memory leak if exception thrown

Use a `std::vector` instead :-)